pkg/controller/utils/datadog/metrics_forwarder.go

```diff
@@ -0,0 +1,538 @@
+	agentEventName              = "Agent"
+	clusteragentName            = "clusteragent"
+	clusteragentEventName       = "Cluster Agent"
+	clustercheckrunnerName      = "clustercheckrunner"
+	clustercheckrunnerEventName = "Cluster Check Runner"
+)
+
+var (
+	// ErrEmptyAPIKey empty APIKey error
+	ErrEmptyAPIKey = errors.New("empty api key")
+	// ErrEmptyAPPKey empty APPKey error
+	ErrEmptyAPPKey = errors.New("empty app key")
+)
+
+var (
+	datadogOperatorSourceType  = "datadog_operator"
+	deploymentCreatedEventType = "created"
+	deploymentDeletedEventType = "deleted"
+)
+
+// metricsForwarder sends metrics directly to Datadog using the public API
+// its lifecycle must be handled by a ForwardersManager
+type metricsForwarder struct {
+	id                  string
+	datadogClient       api.Client
+	k8sClient           client.Client
+	keysHash            uint64
+	retryInterval       time.Duration
+	sendMetricsInterval time.Duration
+	metricsPrefix       string
+	globalTags          []string
+	tags                []string
+	componentEnabled    map[string]bool
+	stop                chan struct{}
+	namespacedName      types.NamespacedName
+	logger              logr.Logger
+	delegator           delegatedAPI
+}
+
+// delegatedAPI is used for testing purpose, it serves for mocking the Datadog API
+type delegatedAPI interface {
+	delegatedSendDeploymentMetric(float64, string, []string) error
+	delegatedSendDeploymentEvent(string, string) error
+	delegatedValidateCreds(string, string) (*api.Client, error)
+}
+
+// newMetricsForwarder returs a new Datadog MetricsForwarder instance
+func newMetricsForwarder(k8sClient client.Client, namespacedName types.NamespacedName) *metricsForwarder {
+	return &metricsForwarder{
+		id:                  namespacedName.String(),
+		k8sClient:           k8sClient,
+		namespacedName:      namespacedName,
+		retryInterval:       defaultMetricsRetryInterval,
+		sendMetricsInterval: defaultSendMetricsInterval,
+		metricsPrefix:       defaultMetricsNamespace,
+		stop:                make(chan struct{}),
+		logger:              log.WithValues("CustomResource.Namespace", namespacedName.Namespace, "CustomResource.Name", namespacedName.Name),
+		componentEnabled: map[string]bool{
+			agentName:              false,
+			clusteragentName:       false,
+			clustercheckrunnerName: false,
+		},
+	}
+}
+
+// start establishes a connection with the Datadog API
+// it starts sending deployment metrics once the connection is validated
+func (mf *metricsForwarder) start(managerChan chan string) {
+	mf.logger.Info("Starting Datadog metrics forwarder")
+
+	// Global tags need to be set only once
+	mf.initGlobalTags()
+
+	// wait.PollInfinite is blocking until mf.connectToDatadogAPI returns true
+	if err := wait.PollInfinite(mf.retryInterval, mf.connectToDatadogAPI); apierrors.IsNotFound(err) {
+		// DatadogAgentDeployment was deleted
+		// inform the ForwardersManager before shutting down the metricsForwarder
+		managerChan <- mf.id
+		return
+	}
```

Comment:
I'm not a big fan of having the metricForwarder responsible to stop himself.
I would prefer to have a `Unregister()` method available in the MetricsForwarders interface
and here: ```func (mf *metricsForwarder) start(stop <-chan struct{}) ```

Reply:
I see your point